### ISSUE TITLE
Add ability to calculate the exact size of the chunks necessary to be downloaded

### ISSUE THREAD

**Issue by RyuzakiKK** [folbricht/desync#270]
## Current situation

We are using Desync to ship system updates, and in order to let the clients estimate the size of an update, we do the following:
- Download the `update.caibx`
- Generate a `current_rootfs.caibx` for the current active partition that will be used as a seed
  - Usually this step is skipped because we save the `.caibx` after applying an update
- Run something like `desync info --seed current_rootfs.caibx update.caibx`
- Parse the `dedup-size-not-in-seed` value
- Reduce that value by 33% to account for the compression rate

## Limitations of this solution

The problem is that `desync info` is only capable of providing information about the _uncompressed_ chunks size that you'll need to download. However, given that we are using compressed chunks, we have no way of knowing exactly how much data will be required to download.

The heuristic of a compression rate of 1.33 places us in the right ballpark, but the real compression rate can varies widely based on how easy/hard those particular chunks can be compressed.

Even if we could get a detailed list of all the chunks that need to be downloaded, it would be unfeasible to send thousands of requests to the server just to ask for their sizes.

### Option 1

One solution could be to extend the `.caibx`/`.caidx` files to include an additional (optional) field about the compressed size for each chunk. However I'm worried this might break backward compatibility with previous Desync versions, or with Casync.

Probably this would require some investigation to ensure we can preserve backward compatibility.

### Option 2

When running `desync make`, or with a separate command, we can store the compressed chunks size in another `.cai*` file. It would be some sort of external file to hold additional optional info about the main `.caibx`/`.caidx` file.

-----

@folbricht what do you think? Do you have other solutions in mind? Before starting spending time to actually design and implement something, I'd like to have your opinion first.

**Comment by folbricht** [comment]
I'm not a fan of option 1. We shouldn't break compatibility for this. Especially since it's not going to be exact/reliable. That's due to stores potentially using different compression levels, a different version of the compression library, or no compression at all.

Do you need a precise number? If you're looking for a better estimate rather than a fixed compression ratio, you could use the chunks you have on the target already to calculate a better one (assuming the new chunks are compressed similarly).

As for option 2, you could achieve something similar today using a simple script. After running `make`, you can run `list` to get a full listing of all chunks in the index, then use that to lookup the file size on disk directly. You could have it produce a listing like so:

```
<chunkid1>,<compressed size1>
<chunkid2>,<compressed size2>
...
```

If it needs to be built-in, perhaps a new sub-command (one which can only be executed against a single local store), somewhat similar to `prune` to produce the necessary information about chunks in it. Similar to the list above or in JSON format? That could then be used as another option on the `info` command to provide the compressed size. But as noted before, this will only be correct if the store that serves these chunks over the network is the same one that the info was extracted from.

**Comment by RyuzakiKK** [comment]
> Do you need a precise number? If you're looking for a better estimate rather than a fixed compression ratio, you could use the chunks you have on the target already to calculate a better one (assuming the new chunks are compressed similarly).

On the target we don't store the compressed chunks, and from our testings the compression ratio of each chunk can varies a lot.
So, we don't _need_ a precise number, but it would be a very nice to have it, if we could.

----

Thanks for the pointers. Once I have some time I'll try to sketch something based on the option 2.

**Comment by folbricht** [comment]
Addressed in https://github.com/folbricht/desync/pull/273
